src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/ModelStateDictionary.cs

```diff
@@ -513,26 +510,159 @@ public void ClearValidationState(string key)
+                {
+                    if (key[closeParanthesesIndex + 1] == '.')
+                    {
+                        // person.Addresses[0].Street
+                        // Create a prefix node for 'Addresses[0].'
+                        index = closeParanthesesIndex + 1;
+                    }
+                    else
+                    {
+                        // person.Addresses[0]Street. Consume the key in entirety
+                        return -1;
+                    }
+                }
+                else
+                {
+                    // person.CarIds[17]. Consume the key in entirety.
+                    return -1;
+                }
+            }
+
+            return index;
+        }
+
+        private ModelStateNode FindNode(string key)
+        {
+            if (key.Length == 0)
+            {
+                return _root;
+            }
+
+            var currentRoot = _root;
+            var previousIndex = 0;
+            int index;
+            while ((index = key.IndexOfAny(Delimiters, previousIndex)) != -1)
+            {
+                index = GetArrayIndexDelimiter(key, index);
+                if (index == -1)
+                {
+                    break;
+                }
+
+                var subKey = new StringSegment(key, previousIndex, index - previousIndex);
+                var next = GetChildNode(currentRoot, subKey);
+                if (next == null)
+                {
+                    return null;
+                }
+
+                currentRoot = next;
+                previousIndex = index + 1;
+            }
+
+            if (previousIndex < key.Length - 1)
+            {
+                var subKey = new StringSegment(key, previousIndex, key.Length - previousIndex);
+                currentRoot = GetChildNode(currentRoot, subKey);
+            }
+
+            return currentRoot;
+        }
+
+        private static ModelStateNode GetChildNode(ModelStateNode currentRoot, StringSegment subKey)
         {
+            if (currentRoot.Children == null)
+            {
+                return null;
+            }
+
+            var searchKey = new KeyValuePair<StringSegment, ModelStateNode>(subKey, value: null);
+            var index = currentRoot.Children.BinarySearch(searchKey, ModelStateNodeComparer.Instance);
+            return index < 0 ? null : currentRoot.Children[index].Value;
+        }
 
+        private static ModelStateNode GetOrAddChildNode(
+            ModelStateNode root,
+            StringSegment subKey,
+            StringSegment key)
+        {
+            ModelStateNode child;
+            if (root.Children == null)
+            {
+                child = new ModelStateNode(root, key);
+                root.Children = new List<KeyValuePair<StringSegment, ModelStateNode>>(1)
+                {
+                    new KeyValuePair<StringSegment, ModelStateNode>(subKey, child)
+                };
```

Comment:
Storing these as KVP makes the list twice as big as necessary. IMO just write a bespoke binary search (~15 loc) instead of this noise.